src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs

```diff
@@ -42,27 +40,30 @@ public abstract class DiagnosticProvider
             /// </summary>
             public abstract Task<IEnumerable<Diagnostic>> GetAllDiagnosticsAsync(Project project, CancellationToken cancellationToken);
 
-            internal async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(
-                FixAllContext fixAllContext)
+            internal async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixAsync(FixAllContext fixAllContext)
             {
                 var result = await GetDocumentDiagnosticsToFixWorkerAsync(fixAllContext).ConfigureAwait(false);
 
                 // Filter out any documents that we don't have any diagnostics for.
                 return result.Where(kvp => !kvp.Value.IsDefaultOrEmpty).ToImmutableDictionary();
-            }
 
-            internal virtual async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixWorkerAsync(
-                FixAllContext fixAllContext)
-            {
-                using (Logger.LogBlock(
-                    FunctionId.CodeFixes_FixAllOccurrencesComputation_Document_Diagnostics,
-                    FixAllLogger.CreateCorrelationLogMessage(fixAllContext.State.CorrelationId),
-                    fixAllContext.CancellationToken))
+                static async Task<ImmutableDictionary<Document, ImmutableArray<Diagnostic>>> GetDocumentDiagnosticsToFixWorkerAsync(FixAllContext fixAllContext)
                 {
-                    return await FixAllContextHelper.GetDocumentDiagnosticsToFixAsync(
-                        fixAllContext,
-                        fixAllContext.ProgressTracker,
-                        (document, cancellationToken) => document.IsGeneratedCode(cancellationToken)).ConfigureAwait(false);
+                    if (fixAllContext.State.DiagnosticProvider is FixAllState.FixMultipleDiagnosticProvider fixMultipleDiagnosticProvider)
```

Comment:
> `fixAllContext.State.DiagnosticProvider`

How/when is this something other than `this`?

📝 This is the main part of the refactoring that doesn't seem like a clear improvement. The original code appeared to be using inheritance in an expected way.

Reply:
> The original code appeared to be using inheritance in an expected way.

Yes, but unfortunately that was using internal virtual APIs, which makes moving code into shared layer not possible. The original approach is reasonable if `IsFixMultiple` is a public API or has more then one true override.

Reply:
@mavasani your thoughts here?

Reply:
> How/when is this something other than this

This is a static local function. Alternatively, we can consider passing in the `this`  argument to the local function, but it seemed easier to just access it off the passed in context object.